src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java

```diff
@@ -47,7 +47,14 @@ public interface ProgrammingExerciseRepository extends JpaRepository<Programming
 
     default ProgrammingExercise findOneByIdElseThrow(final Specification<ProgrammingExercise> specification, long exerciseId) {
         final Specification<ProgrammingExercise> hasIdSpec = (root, query, criteriaBuilder) -> criteriaBuilder.equal(root.get(DomainObject_.ID), exerciseId);
-        return findOne(specification.and(hasIdSpec)).orElseThrow(() -> new EntityNotFoundException("Programming Exercise", exerciseId));
+
+        Optional<ProgrammingExercise> exerciseOptional = findOne(hasIdSpec);
+
+        if (exerciseOptional.isEmpty()) {
+            throw new EntityNotFoundException("Programming Exercise", exerciseId);
+        }
+
+        return findAll(specification.and(hasIdSpec)).getFirst();
```

Comment:
It actually doesn't. If you have a query expecting one element or an optional element, it fails if there is more than one row returned.

Reply:
I'm also not happy about this change. If the optional part is an issue and we definitely need two queries, the first query could only return a boolean whether the exercise exists or not (because this avoids EAGER fetches).

I guess we should implement this particular query with the approach in #8607

Reply:
Ah, I realise why this is the case: Otherwise, Spring Boot has no chance of detecting whether multiple results were found.
I see that you wrote something similar, I just didn't process the "only" in your sentence. My bad.

Thus, I would not override it.
However, maybe we can create an internal findOne implementation that doesn't override, but gets used by the RepositoryImpl.

This way we solve the issue, but only where we fetch **by ID**. If we fetch only one single record by some random query, I would also want to make sure that there are no duplicate entries when I only expect one.

Reply:
I know, but if the result would get limited to one before, there is always max one.